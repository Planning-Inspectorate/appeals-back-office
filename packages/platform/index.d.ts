import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
