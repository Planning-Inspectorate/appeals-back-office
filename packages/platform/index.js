export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
