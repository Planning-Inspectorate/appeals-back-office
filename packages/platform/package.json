{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
