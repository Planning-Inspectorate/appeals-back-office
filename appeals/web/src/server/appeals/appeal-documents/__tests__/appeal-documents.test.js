import nock from 'nock';
import supertest from 'supertest';
import { parseHtml } from '@pins/platform';
import { createTestEnvironment } from '#testing/index.js';
import {
	appellantCaseDataNotValidated,
	documentFileVersionsInfo
} from '#testing/app/fixtures/referencedata.js';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const validAppealId = 1;
const invalidAppealId = 2;
const validFolderId = 1;
const invalidFolderId = 2;
const documentId = '0e4ce48f-2d67-4659-9082-e80a15182386';
const validFolders = [
	{ folderId: validFolderId, path: 'appellantCase/newSupportingDocuments', caseId: validAppealId }
];

const getControllerEndpoint = (
	/** @type {number} */ appealId,
	/** @type {number} */ folderId,
	/** @type {string|undefined} */ documentId
) => {
	let baseUrl = `/appeals-service/appeal-details/${appealId}/appellant-case/add-documents/${folderId}`;
	if (documentId) {
		baseUrl += `/${documentId}`;
	}
	return baseUrl;
};

describe('documents upload', () => {
	beforeEach(() => {
		installMockApi();
		nock('http://test/')
			.get('/appeals/1/appellant-cases/0')
			.reply(200, appellantCaseDataNotValidated);
	});
	afterEach(teardown);

	it('should return 404 if appeal ID is not found', async () => {
		nock('http://test/').get(`/appeals/${invalidAppealId}`).reply(404);
		nock('http://test/').get(`/appeals/${invalidAppealId}/document-folders/1`).reply(404);

		const response = await request.get(getControllerEndpoint(invalidAppealId, invalidFolderId));

		expect(response.statusCode).toBe(404);
	});

	it('should return 404 if folder ID is not found', async () => {
		nock('http://test/').get(`/appeals/${validAppealId}`).reply(200, { id: validAppealId });
		nock('http://test/')
			.get(`/appeals/${validAppealId}/document-folders/1`)
			.reply(200, validFolders);

		const response = await request.get(getControllerEndpoint(validAppealId, invalidFolderId));
		expect(response.status).toBe(404);
	});

	it('should return 404 if document ID is not found', async () => {
		nock('http://test/').get(`/appeals/${validAppealId}`).reply(200, { id: validAppealId });
		nock('http://test/')
			.get(`/appeals/${validAppealId}/document-folders/1`)
			.reply(200, validFolders[0]);
		nock('http://test/').get(`/appeals/${invalidAppealId}/documents/${documentId}`).reply(404);

		const response = await request.get(
			getControllerEndpoint(validAppealId, validFolderId, documentId)
		);
		expect(response.status).toBe(404);
	});

	it('should render upload form if appeal ID and folder ID are found', async () => {
		nock('http://test/').get(`/appeals/${validAppealId}`).reply(200, { id: validAppealId });
		nock('http://test/')
			.get(`/appeals/${validAppealId}/document-folders/1`)
			.reply(200, validFolders[0]);

		const response = await request.get(getControllerEndpoint(validAppealId, validFolderId));
		expect(response.status).toBe(200);

		const html = parseHtml(response.text);
		expect(html.innerHTML).toMatchSnapshot();

		const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

		expect(unprettifiedElement.innerHTML).toContain('Upload documents</h1>');
		expect(unprettifiedElement.innerHTML).toContain('<form method="POST"');
		expect(unprettifiedElement.innerHTML).toContain('Choose file</button>');
		expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
	});

	it('should render upload form if appeal ID, folder ID and document ID are found', async () => {
		nock('http://test/').get(`/appeals/${validAppealId}`).reply(200, { id: validAppealId });
		nock('http://test/')
			.get(`/appeals/${validAppealId}/document-folders/1`)
			.reply(200, validFolders[0]);
		nock('http://test/')
			.get(`/appeals/${validAppealId}/documents/${documentId}`)
			.reply(200, {
				caseId: validAppealId,
				id: documentId,
				name: 'Doc A',
				latestDocumentVersion: {
					fileName: 'A',
					documentType: 'docs',
					version: 1
				}
			});
		nock('http://test/')
			.get(`/appeals/${validAppealId}/documents/${documentId}/versions`)
			.reply(200, documentFileVersionsInfo);

		const response = await request.get(
			getControllerEndpoint(validAppealId, validFolderId, documentId)
		);
		expect(response.status).toBe(200);

		const html = parseHtml(response.text);
		expect(html.innerHTML).toMatchSnapshot();

		const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

		expect(unprettifiedElement.innerHTML).toContain('Upload an updated document</h1>');
		expect(unprettifiedElement.innerHTML).toContain('<form method="POST"');
		expect(unprettifiedElement.innerHTML).toContain('Choose file</button>');
		expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
	});

	it('should render appeal ID and folder ID as data attributes', async () => {
		nock('http://test/').get(`/appeals/${validAppealId}`).reply(200, { id: validAppealId });
		nock('http://test/')
			.get(`/appeals/${validAppealId}/document-folders/1`)
			.reply(200, validFolders[0]);

		const response = await request.get(getControllerEndpoint(validAppealId, validFolderId));
		const html = parseHtml(response.text);

		const dataAttributes = processAttrs(html.querySelector('.pins-file-upload')?.attributes);
		// @ts-ignore
		expect(dataAttributes['data-case-id']).toEqual(validAppealId.toString());
		// @ts-ignore
		expect(dataAttributes['data-folder-id']).toEqual(validFolderId.toString());
		// @ts-ignore
		expect(dataAttributes['data-document-id']).toBeUndefined();
	});

	it('should render all necessary metadata attributes', async () => {
		nock('http://test/').get(`/appeals/${validAppealId}`).reply(200, { id: validAppealId });
		nock('http://test/')
			.get(`/appeals/${validAppealId}/document-folders/1`)
			.reply(200, validFolders[0]);
		nock('http://test/')
			.get(`/appeals/${validAppealId}/documents/${documentId}`)
			.reply(200, { latestDocumentVersion: {} });
		nock('http://test/')
			.get(`/appeals/${validAppealId}/documents/${documentId}/versions`)
			.reply(200, documentFileVersionsInfo);

		const response = await request.get(
			getControllerEndpoint(validAppealId, validFolderId, documentId)
		);
		const html = parseHtml(response.text);

		const dataAttributes = processAttrs(html.querySelector('.pins-file-upload')?.attributes);
		// @ts-ignore
		expect(dataAttributes['data-case-id']).toEqual(validAppealId.toString());
		// @ts-ignore
		expect(dataAttributes['data-folder-id']).toEqual(validFolderId.toString());
		// @ts-ignore
		expect(dataAttributes['data-document-id']).toEqual(documentId);
		// @ts-ignore
		expect(dataAttributes['data-document-stage']).toEqual(validFolders[0].path.split('/')[0]);
		// @ts-ignore
		expect(dataAttributes['data-document-type']).toEqual(validFolders[0].path.split('/')[1]);
	});

	it('should render blob host and container', async () => {
		nock('http://test/').get(`/appeals/${validAppealId}`).reply(200, { id: validAppealId });
		nock('http://test/')
			.get(`/appeals/${validAppealId}/document-folders/1`)
			.reply(200, validFolders[0]);

		const response = await request.get(getControllerEndpoint(validAppealId, validFolderId));
		const html = parseHtml(response.text);

		const dataAttributes = processAttrs(html.querySelector('.pins-file-upload')?.attributes);
		// @ts-ignore
		expect(dataAttributes['data-blob-storage-host']).not.toBe(null);
		// @ts-ignore
		expect(dataAttributes['data-blob-storage-container']).not.toBe(null);
	});
});

const processAttrs = (/** @type {NamedNodeMap | undefined} */ attrs) => {
	const items = {};
	if (attrs) {
		for (const item of Object.keys(attrs)) {
			if (item.indexOf('data') === 0) {
				// @ts-ignore
				items[item] = attrs[item];
			}
		}
	}

	return items;
};
