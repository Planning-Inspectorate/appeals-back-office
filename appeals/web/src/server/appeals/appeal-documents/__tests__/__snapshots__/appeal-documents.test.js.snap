// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appeal-documents upload should render upload form if appeal ID and folder ID are found 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds top-errors-hook"></div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">Upload your other new supporting documents</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-full">
            <div class="govuk-grid-row pins-file-upload" data-next-page-url="/appeals-service/appeal-details/1/appellant-case/add-document-details/1"
            data-form-id="1" data-case-id="1" data-case-reference="APP/Q9999/D/21/351062"
            data-folder-id="1" data-document-type="otherNewDocuments" data-document-stage="appellantCase"
            data-blob-storage-host="https://127.0.0.1:10000/devstoreaccount1" data-blob-storage-container="document-service-uploads"
            data-document-title="other new supporting documents" data-allowed-types="application/pdf,application/msword,application/vnd.openxmlformats-officedocument.wordprocessingml.document,application/vnd.ms-powerpoint,application/vnd.openxmlformats-officedocument.presentationml.presentation,application/vnd.ms-excel,application/vnd.openxmlformats-officedocument.spreadsheetml.sheet,application/vnd.ms-outlook,image/jpeg,image/jpeg,video/mpeg,audio/mpeg,video/mp4,video/quicktime,image/png,image/tiff,image/tiff"
            data-formatted-allowed-types="PDF, DOC, DOCX, PPT, PPTX, XLS, XLSX, MSG, JPG, JPEG, MPEG, MP3, MP4, MOV, PNG, TIF or TIFF">
                <form method="POST" novalidate="novalidate" class="govuk-grid-column-two-thirds"
                data-filenames-in-folder="">
                    <div class="pins-file-upload__container">
                        <div class="govuk-body pins-file-upload__instructions"><span>Each file must be a PDF, DOC, DOCX, PPT, PPTX, XLS, XLSX, MSG, JPG, JPEG, MPEG, MP3, MP4, MOV, PNG, TIF or TIFF and smaller than 25MB.</span><span>The total size of your uploaded files must be smaller than 1GB.</span>
                        </div>
                        <div class="middle-errors-hook">
                            <h2 class="pins-file-upload__container-title">Upload your other new supporting documents</h2>
                            <div class="pins-file-upload__upload">
                                <input class="display--none" id="upload-file-1" accept="application/pdf,application/msword,application/vnd.openxmlformats-officedocument.wordprocessingml.document,application/vnd.ms-powerpoint,application/vnd.openxmlformats-officedocument.presentationml.presentation,application/vnd.ms-excel,application/vnd.openxmlformats-officedocument.spreadsheetml.sheet,application/vnd.ms-outlook,image/jpeg,image/jpeg,video/mpeg,audio/mpeg,video/mp4,video/quicktime,image/png,image/tiff,image/tiff"
                                type="file" name="files" value="Choose file" aria-controls="file-list-1"
                                multiple>
                                <div>
                                    <button type='button' class="pins-file-upload__button govuk-button--secondary"
                                    data-cy="upload-file-button" id="upload-file-button-1">Choose files</button><span class='govuk-body pins-file-upload__dropzone-text'>or drop files</span>
                                </div>
                            </div>
                            <div id="file-list-1">
                                <h2 class="display--sr-only" id="file-list-title-1">List of files to upload</h2>
                                <ul class="pins-file-upload__files-rows" aria-describedby="file-list-title-1"
                                aria-live="polite"></ul>
                            </div>
                        </div>
                    </div>
                    <div class="display--flex">
                        <button type="submit" class="govuk-button pins-file-upload__submit" data-module="govuk-button"
                        id="submit-button-1">Continue</button>
                        <div class="progress-hook"></div>
                    </div>
                </form>
            </div>
        </div>
    </div>
</main>"
`;

exports[`appeal-documents upload should render upload form if appeal ID, folder ID and document ID are found 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds top-errors-hook"></div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">Upload an updated document</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-full">
            <div class="govuk-grid-row pins-file-upload" data-next-page-url="/appeals-service/appeal-details/1/appellant-case/add-document-details/1/0e4ce48f-2d67-4659-9082-e80a15182386"
            data-form-id="1" data-case-id="1" data-case-reference="APP/Q9999/D/21/351062"
            data-folder-id="1" data-document-type="otherNewDocuments" data-document-stage="appellantCase"
            data-blob-storage-host="https://127.0.0.1:10000/devstoreaccount1" data-blob-storage-container="document-service-uploads"
            data-document-id="0e4ce48f-2d67-4659-9082-e80a15182386" data-allowed-types="application/pdf"
            data-formatted-allowed-types="PDF">
                <form method="POST" novalidate="novalidate" class="govuk-grid-column-two-thirds"
                data-filenames-in-folder="">
                    <div class="pins-file-upload__container">
                        <div class="govuk-body pins-file-upload__instructions"><span>The file must be a PDF and smaller than 25MB.</span>
                        </div>
                        <div class="middle-errors-hook">
                            <h2 class="pins-file-upload__container-title">Upload an updated document</h2>
                            <div class="pins-file-upload__upload">
                                <input class="display--none" id="upload-file-1" accept="application/pdf"
                                type="file" name="files" value="Choose file" aria-controls="file-list-1">
                                <div>
                                    <button type='button' class="pins-file-upload__button govuk-button--secondary"
                                    data-cy="upload-file-button" id="upload-file-button-1">Choose file</button><span class='govuk-body pins-file-upload__dropzone-text'>or drop file</span>
                                </div>
                            </div>
                            <div id="file-list-1">
                                <h2 class="display--sr-only" id="file-list-title-1">List of files to upload</h2>
                                <ul class="pins-file-upload__files-rows" aria-describedby="file-list-title-1"
                                aria-live="polite"></ul>
                            </div>
                        </div>
                    </div>
                    <div class="display--flex">
                        <button type="submit" class="govuk-button pins-file-upload__submit" data-module="govuk-button"
                        id="submit-button-1">Continue</button>
                        <div class="progress-hook"></div>
                    </div>
                </form>
            </div>
        </div>
    </div>
</main>"
`;
